### PR TITLE
[DNM] storage: Store region's max_read_ts and report it to TiDB in prewrite response

### PR DESCRIPTION
**Pull request by MyonKeminta**
**Do not merge** : This PR is totally experimental, used to find out the effect of avoiding TiDB allocating commit_ts from PD.

## What have you changed? (mandatory)

Recorded maximum timestamp of transactional reads of each region, and return it in prewrite response to make TiDB able to calculate a commit_ts.

Requires #4402 and #4382 
RFC at https://github.com/tikv/rfcs/pull/25

## What are the type of the changes? (mandatory)

- Improvement (change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

By manual tests

## Does this PR affect documentation (docs) update? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

### CODE REVIEW

File: src/storage/inspector.rs [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+                    region_id: ctx.region().get_id(),
+                },
+            };
+            self.scheduler.schedule(event).unwrap();
+        }
+    }
+}
+
+// region_id -> (epoch_version, max_read_ts)
+type TsMap = HashMap<(u64, u64), Arc<AtomicU64>>;
```
Comment (siddontang) [inline src/storage/inspector.rs]: I guess using one map may cause too many contentions and reduce performance, maybe we should use arrays of map like Vec<Arc<RwLock<TsMap>> and hash the region id to a map at first.

Comment (MyonKeminta) [reply src/storage/inspector.rs]: Ok. I'm getting into some problems about TiDB's code now... and after that, I'll try more different implementations here.

**Comment by MyonKeminta** [comment]
The name "inspector" is not suitable for the module that keeps region's max_read_ts. Is there any one who have better idea about its name?

**Comment by MyonKeminta** [comment]
This optimization is found incompatible with tidb-binlog and Raft learner backup. Close it. Reopen when it's ok to continue.

**Pull request closed without merging**